### ISSUE TITLE
k8s: Cross-host pod connections not properly rendered or missing

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/scope#1404]
Pod connections (no services nor kube-proxy involved) across hosts in kubernetes are not properly reflected in the UI. 

Following a modified example from https://github.com/kelseyhightower/craft-kubernetes-workshop , I am using an nginx `frontend` pod (IP 10.152.1.6) which uses persistent connections to loadbalance requests to some `hello` pod backends with IPs (10.152.1.5, 10.152.0.3 and 10.152.4.5). 

nginx is using this configuration:

```
upstream hello {                                                                                                                                            
    server 10.152.1.5;                                                                                                                                      
    server 10.152.0.3;                                                                                                                                      
    server 10.152.4.5;                                                                                                                                      
    keepalive 100;                                                                                                                                            
}                                                                                                                                                           
                                                                                                                                                            
upstream auth {                                                                                                                                             
    server auth.default.svc.cluster.local;                                                                                                                  
    keepalive 2;                                                                                                                                            
}                                                                                                                                                           
                                                                                                                                                            
server {                                                                                                                                                    
    listen 443;                                                                                                                                             
    ssl    on;                                                                                                                                              
                                                                                                                                                            
    ssl_certificate     /etc/tls/cert.pem;                                                                                                                  
    ssl_certificate_key /etc/tls/key.pem;                                                                                                                   
                                                                                                                                                            
    proxy_http_version 1.1;                                                                                                                                 
    proxy_set_header Connection "";                                                                                                                         
                                                                                                                                                            
    location / {                                                                                                                                            
        proxy_pass http://hello;                                                                                                                            
    }                                                                                                                                                       
                                                                                                                                                            
    location /login {                                                                                                                                       
        proxy_pass http://auth;                                                                                                                             
    }                                                                                                                                                       
}
```

The frontend shows connections to all the pods:

```
$ netstat -tpn | grep ESTABLISHED  | grep 10.152.0.3                                                                                                         
tcp        0      0 10.152.1.6:60463        10.152.0.3:80           ESTABLISHED -                                                                           
$ netstat -tpn | grep ESTABLISHED  | grep 10.152.1.5                                                                                                         
tcp        0      0 10.152.1.6:42595        10.152.1.5:80           ESTABLISHED -                                                                           
$ netstat -tpn | grep ESTABLISHED  | grep 10.152.4.5                                                                                                         
tcp        0      0 10.152.1.6:42416        10.152.4.5:80           ESTABLISHED -                                                                           
```

Similarly, the backends show connections from the frontend. For instance:

```
/ # netstat -pnt                                                                                                                                            
Active Internet connections (w/o servers)                                                                                                                   
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name                                                            
tcp        0      0 ::ffff:10.152.0.3:80    ::ffff:10.152.1.6:60463 ESTABLISHED 1/hello                                                                     
/ #                                                                                                                                                         
Connected
```

```
/ # netstat -npt                                                                                                                                            
Active Internet connections (w/o servers)                                                                                                                   
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name                                                            
tcp        0      0 ::ffff:10.152.1.5:80    ::ffff:10.152.1.6:42595 ESTABLISHED 1/hello                                                                     
/ #                                                                                                                                                         
```

Pods 10.152.1.6 and 10.152.1.5 are co-located in the same host, and their container connections are properly rendered:

<img width="1747" alt="screen shot 2016-04-29 at 00 05 58" src="https://cloud.githubusercontent.com/assets/2362916/14904163/47359e60-0d9e-11e6-8b50-01ed5913d68d.png">


However that's not the case for cross-host pod connections. There are three problems:

1. One end is attributed to the `/pause` container, which is tagged as a system container and consequently is not displayed in the normal view. Scope has the PID of the destination process since the connection is persistent so it should be able to attribute it to right container (Related https://github.com/weaveworks/scope/issues/1394 )

2. The other end is attributed to _The Internet_. Although each kubernetes node uses a different network for its pod IPs the app should know how to join them together.

    <img width="1746" alt="screen shot 2016-04-29 at 00 14 23" src="https://cloud.githubusercontent.com/assets/2362916/14904314/5ea2dd3c-0d9f-11e6-8878-f7e94a83189e.png">

3. Some connections aren't rendered, for instance: `10.152.1.6:60463  <-> 10.152.0.3:80`.
    <img width="1747" alt="screen shot 2016-04-29 at 00 27 19" src="https://cloud.githubusercontent.com/assets/2362916/14904602/2beb28b6-0da1-11e6-9712-5c7c5aac7ab0.png">


    But the report seems to have enough information to render it (although I wonder why the connection is conntracked instead of obtained from `/proc`).

    <img width="404" alt="screen shot 2016-04-29 at 00 25 12" src="https://cloud.githubusercontent.com/assets/2362916/14904562/e5944370-0da0-11e6-8442-fafb9f24b551.png">
    <img width="414" alt="screen shot 2016-04-29 at 00 25 01" src="https://cloud.githubusercontent.com/assets/2362916/14904561/e5933bba-0da0-11e6-9aba-8fb85e3b6613.png">
    <img width="408" alt="screen shot 2016-04-29 at 00 24 44" src="https://cloud.githubusercontent.com/assets/2362916/14904563/e5959090-0da0-11e6-88b4-f67fec60a533.png">
    <img width="413" alt="screen shot 2016-04-29 at 00 24 23" src="https://cloud.githubusercontent.com/assets/2362916/14904560/e5926474-0da0-11e6-8427-0da671e029c6.png">


It seems like connections figured out from `/proc` aren't properly merged with the ones figured out from conntrack (the first should have higher precedence since they provide PIDs).

Report: [report_crosshost_pod_connections.json.gz](https://github.com/weaveworks/scope/files/241579/report_crosshost_pod_connections.json.gz)

**Comment by kelseyhightower** [comment]
@2opremio This issue captures everything perfectly.

**Comment by 2opremio** [comment]
> This issue captures everything perfectly.

Great, we will try to get this fixed soon, hopefully next week.

**Comment by tomwilkie** [comment]
Fixed by #1410
